Overhaul the labels page to reflect current GH labeling practice (#930)
python/devguide: triage/labels.rst

only to one. Below is a possibly inexhaustive list, but it should get
you going. For a full list, see `here <https://github.com/python/cpython/issues/labels>`_.

General purpose labels

> ======================

Reviewer: We could re-add it as a ref target label to the closest equivalent, the `Type Labels` section which is also at the top of the main labels section (and is the most general). I've made a suggestion accordingly.

> ======================

Author: Is there anything linking to that in the devguide?  I wouldn't worry too much about external links here, since these labels are somewhat specific to CPython.  In addition, even if the link is broken, it should be easy enough to find the relevant information in the page.

This also applies to other comments below, especially when there was no explicit label before the section.

> ======================

Reviewer: IMO, we should definitely preserve existing ref target labels, moving them to the new location of the information (or closest equivalent) as needed, unless the relevant information was completely removed with no even approximate equivalent. Their express purpose and value is in being stable, relocatable references to information not only internally, but also externally through both intersphinx and URI fragments. Section headers are somewhat less critical, but still, IMO, helpful. My typical strategy nowadays is to just add ref target labels for all sections, namespaced by a short identifier for the page, so it is easy and painless to link to any section whether internally, through intersphinx or externally without having to first go and modify the source.

To me, internal links are not really a concern since it is relatively trivial to synchronously update them, whereas external links are much more important, as they could be both in other docs sites and all over the web in numerous places that we might not even think about, much less be able to update—Mailing list archives, Discourse, forums, Stack Overflow, blogs, websites, Reddit, GitHub—that would otherwise break and linkrot. The cost is definitely something to consider if it involves multiple lines of non-trivial content, and appears in the rendered output, but a single extra short ref target label seems like a very small cost relative to the potential breakage.

> Type labels

Reviewer: ```suggestion
.. _general-purpose-labels:

Type labels
```

> Type labels
> ===========

Reviewer 2: The Type section things were moved here, also move the anchor?

```suggestion
.. _Type:

Type labels
===========
```

These labels are used to specify the type of issue:

* :gh-label:`type-bug`: for unexpected behaviors, bugs, or exceptions
  (not hard crashes).
* :gh-label:`type-crash`: for hard crashes of the interpreter, possibly with a
  core dump.
* :gh-label:`type-feature`: for feature requests or enhancements.
  The `Ideas Discourse category`_ can be used to discuss enhancements
  before filing an issue.
* :gh-label:`type-security`: for security issues.
  See also `Reporting security issues in Python`_.

Since most PRs have an associated issue, it is not necessary to apply these
labels to PRs.

> 

Author: Is there any consensus on what to do before documenting it?

> Component labels
> ================

Reviewer 2: Move `#component` anchor here:

```suggestion
.. _Component:

Component labels
================
```

  and :cpy-file:`Parser` dirs (written mostly in C).
* :gh-label:`docs`: for documentation in the :cpy-file:`Doc` directory
  (written in :ref:`reStructuredText <markup>`), docstrings, and code comments.
* :gh-label:`tests`: for tests in the :cpy-file:`Lib/test` directory
  (written in Python) and other changes related to tests, ``unittest``,
  or ``doctests``.

Expert labels
=============

These labels are used to specify the area of expertise required to address
the issue/PR.  This includes both specific modules/packages and generic
interest areas.

Adding these labels is also a way to notify the relevant experts, since
they are encouraged to subscribe to them.  Depending on the label,
this might also automatically add the issue to a GitHub project.

You can see the `full list of expert labels on GitHub
<https://github.com/python/cpython/labels?q=expert>`_.

OS labels
=========

These labels are used to specify which operating systems are affected.
Since most issues either affect all systems or are specific to Unix,
the only available labels are :gh-label:`OS-windows`, :gh-label:`OS-mac`,
and :gh-label:`OS-freebsd`.

> Version labels
> ==============

Reviewer 2: ```suggestion
.. Versions:

Version labels
==============
```

Reviewer: Fixed a typo in your suggestion (missing leading `_`)

These labels are used to indicate which versions of Python are affected.

> The available version labels (with the form :samp:`3.{x}`) are updated

Reviewer 2: I think 3.x is fine here and much more common. I can't recall seeing, or find with a quick search, any Python 3.y.

Things like PEP 440 or the SemVer spec are different, when discussing X.Y.Z and their relationship, but here the X is the first unknown. (And the PEP has much more N.N than X.Y.)

> The available version labels (with the form :samp:`3.{x}`) are updated

Reviewer: In that case, then,

```suggestion
The available version labels (with the form :samp:`3.{N}`) are updated
```

> The available version labels (with the form :samp:`3.{x}`) are updated

Reviewer: Just to clarify, the usage I've typically seen for `3.x` is _any 3.* version_, _all 3.* versions_, etc., whereas here the labels have a _specific_ `3.{N}` version, with the meaning of the character being `{}` (i.e. `.{N}`) rather than `*` (i.e. `.x`).

whenever new major releases are created or retired.

See also :ref:`the branch status page <branchstatus>`
for a list of active branches.

> Other labels
> ============

Reviewer 2: These anchors were deleted, their labels moved here. Shall we add them back to avoid linkrot?

```suggestion
.. _Keywords:
.. _Other:
.. _Priority:

Other labels
============
```

* :gh-label:`triaged`: for issue has been accepted as valid by a triager.
* :gh-label:`easy`: for issues that are considered easy.
* :gh-label:`build`/:gh-label:`performance`: for issues related respectively
  to the build process and performances.

> * :gh-label:`release-blocker`/:gh-label:`deferred-blocker`: for issues/PRs
>   that, unless fixed, will hold the current or next release respectively.

Reviewer: Actually, I see that was discussed in the previous section was adapted from, so I've attempted to condense the key points from there and include them here:

```suggestion
* :gh-label:`release-blocker`/:gh-label:`deferred-blocker`: for issues/PRs
  that, unless fixed, will hold the current or next release respectively.
  These should generally only be apply by or at the direction of the
  :ref:`release manager of the relevant branch <branchstatus>`.
  Triagers can tag them on an issue comment using :samp:`@{username}`.
```

> * :gh-label:`release-blocker`/:gh-label:`deferred-blocker`: for issues/PRs
>   that, unless fixed, will hold the current or next release respectively.

Reviewer: This modifies the workflow to be the one @ezio-melotti [suggests](https://github.com/python/devguide/pull/930#discussion_r1057070577), though it should also have at least a couple additional lines describing specifically what generally qualifies as a possible release blocker, so triagers understand how (and how not) to apply the label.

```suggestion
* :gh-label:`release-blocker`/:gh-label:`deferred-blocker`: for issues/PRs
  that, unless fixed, will hold the current or next release respectively.
  Triagers may set these labels, and the :ref:`branch's release manager <branchstatus>`
  will review them and determine if they indeed quality,
  removing or retaining the label as appropriate.
```

>   that, unless fixed, will hold the current or next release respectively.

Author: Fixed a typo in your suggestion (*quality* -> *qualify*).

* :gh-label:`pending`: for issues/PRs that will be closed unless further
  feedback is provided.
* :gh-label:`stale`: for issues/PRs that have been inactive for a while.
* :gh-label:`sprint`: used for easier filtering of issues/PRs being worked on
  during official sprints.

> Labels specific to PRs
> ======================

Reviewer 2: The "GitHub Labels for PRs" section below was deleted, move its anchor here?

```suggestion
.. GitHub Labels for PRs:

Labels specific to PRs
======================
```

Reviewer: (Fixed a typo in your comment)

| Interpreter Core  | The interpreter core.                                |
|                   | The built-in objects in `Objects`_, the `Python`_,   |
|                   | `Grammar`_ and `Parser`_ dirs.                       |
+-------------------+------------------------------------------------------+
| Library (Lib)     | Python modules in Lib_.                              |
+-------------------+------------------------------------------------------+
| Tests             | The unittest framework in `Lib/unittest`_            |
|                   | The doctest framework `Lib/doctest.py`_.             |
|                   | The CPython tests in `Lib/test`_.                    |
|                   | The test runner in `Lib/test/regrtest.py`_.          |
|                   | The test support utilities in `Lib/test/support`_.   |
+-------------------+------------------------------------------------------+

Versions
--------
The known versions of Python that the issue affects and should be fixed for.

Thus if an issue for a new feature is assigned for e.g., Python 3.8 but is not
applied before Python 3.8.0 is released, this label should be updated to say

> ``python-3.9`` as the version and drop ``python-3.8``.

Author: This has been replaced by the `Version labels` section above.

| normal           | The default value for most issues filed.               |
+------------------+--------------------------------------------------------+
| deferred blocker | The issue will not hold up the next release, *n*. It   |
|                  | will be promoted to a *release blocker* for the        |
|                  | following release, *n+1*.                              |
+------------------+--------------------------------------------------------+
| release blocker  | The issue **must** be fixed before *any* release is    |
|                  | made, e.g., will block the next release even if it is  |
|                  | an alpha release.                                      |
+------------------+--------------------------------------------------------+

As a guideline, whether a bug is a *release blocker* for the current
:ref:`release schedule <branchstatus>` is decided by the release manager.
Triagers may recommend this priority and should notify the release manager by
tagging them in a comment using ``@username``. If needed, consult the
:ref:`release schedule <branchstatus>` and the release's associated PEP for the

> release manager's name.

Author: This was also adapted from BPO and it's now covered in the `Other labels` section.

Reviewer: It would seem to be very helpful to port over the key points of the paragraph at the bottom, though, to make clear that the RM generally sets these labels, not triagers (or even core devs)? I recall this being a point of confusion at least a couple times in the past, when a triager set this label not knowing that it was the RM's call instead. I've added a suggestion above implementing this.

Author: I was under the impression that triagers could set these labels to bring the issue up to the attention of the RMs if they think it's a blocker.  Since the RMs should be subscribed to the labels and follow the "Release and deferred blockers" project, they should get notified and can then decide whether it was a good call or not, and possibly remove the label.

This also solves the issue of having to figure out who the RMs are, pinging them manually, and risking that the ping gets lost among other notifications.

Perhaps we should ask the RMs if this workflows is preferable or if they prefer to be pinged manually as the removed docs suggests.

P.S. apparently we have a @python/release-managers-in-development-maintenance-and-security-mode team that could be pinged instead of looking up the RM name.

Reviewer: That's a great point. The current text implies (but does not explicitly state) that triagers should only _ping_ the RM in a comment _recommending_ this priority (with no explicit guidance for core devs), while the RM should apply it, and I was also led to believe the same by one of Pablo's Discord comments in response to a triager proposing a release blocker on there, that "all release blockers should be validated by the RM", which [my previous suggestion](https://github.com/python/devguide/pull/930#discussion_r1056997730) reflects.

However, the does sound like a potentially simpler and more efficient workflow for triagers and RMs alike. However, it relies on triagers being able to have a reasonable idea of what might qualify as a release blocker, which in turn requires us to provide reasonably clear and specific guidance to them here, as well as the RMs being comfortable with that. In that case, I've made [a followup suggestion](https://github.com/python/devguide/pull/930#discussion_r1057294582) implementing this alternative instead.

Keywords
--------
Various informational flags about the issue. Multiple values are possible.

+---------------+------------------------------------------------------------+
|    Keyword    |                        Description                         |
+===============+============================================================+
| easy          | Fixing the issue should not take longer than a day for     |
|               | someone new to contributing to Python to solve.            |

> +---------------+------------------------------------------------------------+

Author: Ditto.

Nosy List
---------
A list of people who may be interested in an issue.

This used to be a feature of the old issue tracker. On GitHub issues the
same effect is achieved by tagging people in a comment using ``@username``.

It is acceptable to tag someone to if you think the issue should be brought to
their attention. Use the :ref:`experts` to know who wants to be added to the
nosy list for issues targeting specific areas.

If you want to subscribe yourself to an issue, click the *🔔 Subscribe*
button in the sidebar. Similarly, if you were tagged by somebody else but
decided this issue is not for you, you might click the *🔕 Unsubscribe*

> button in the sidebar.

Author: Part of this has been adapted and moved to the `issue-tracker.rst` page.

Reviewer: Seems like most of the still relevant material is retained, but it is missing a bit of prior context about _how_ to tag people, especially given the audience there may include beginners not familiar with using GitHub. I'll make a suggestion there.

Assignees
---------
Who is expected to take the next step in resolving the issue.

It is acceptable to assign an issue to someone if the issue cannot move
forward without their help, e.g., they need to make a technical decision to
allow the issue to move forward. Also consult the :ref:`experts` as certain
stdlib modules should always be assigned to a specific person.

Note that in order to assign an issue to someone, that person **must** be

> a team member, likely a Triager or a core developer.

Author: This is briefly covered in the `issue-tracker.rst` page.

Reviewer: It might still be helpful, though, to still provide some guidance about what the `Assignees` field should be used for, presumably on the `Triaging and Issue` page, which the section added to the Issue Tracker page references readers toward (but I can't find any more than a similarly brief mention of it either place). I can't find the details currently discussed here anywhere else, and this seems particularly germane given even among experienced triagers and core devs, there doesn't appear to be a lot of clarity or consistency regarding how to best use this field

Reviewer: Seems like this is still missing; the mention on the `issue-tracker` page only states that regular users should not need to touch it and refers the reader to this page for more information, on which there now is none. I would suggest just restoring the text as-is for now and we can iterate on it in the future as necessary, as it is a net regression otherwise.

Author: I now moved in to `triaging.rst`.  I left a note about documenting the other fields too.

Reviewer: Great, thanks!

Dependencies
------------
The issue requires the listed issue(s) to be resolved first before it can move
forward. This is achieved using checkbox lists in the initial issue description
comment. Long story short, if you add this::

    - [x] #739
    - [ ] https://github.com/octo-org/octo-repo/issues/740
    - [ ] Add delight to the experience when all tasks are complete :tada:

then those will become sub-tasks on the given issue. Moreover, GitHub will
automatically mark a task as complete if the other referenced issue is
closed.

More details in the `official GitHub documentation
<https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/about-task-lists>`_.

Superseder
----------
The issue is a duplicate of the listed issue(s). To make GitHub mark

> an issue as duplicate, write "Duplicate of #xxxx" in a comment.

Author: These have been adapted and moved to the `issue-tracker.rst` page.

Reviewer 2: For these that have been moved to another page, is it worth preserving the `#anchors` by putting them at the bottom, like we have on these other pages?

https://devguide.python.org/documentation/start-documenting/#style-guide
https://devguide.python.org/documentation/start-documenting/#translating

```rst
Style Guide
===========

Moved to :doc:`style-guide`

Translating
===========

Moved to :doc:`translating`
```

https://devguide.python.org/#status-of-python-branches

```rst
Status of Python branches
-------------------------

Moved to :ref:`versions`
```

Status
------

+---------------+------------------------------------------------------------+
|    Status     |                        Description                         |
+===============+============================================================+
| open          | Issue is not resolved.                                     |
+---------------+------------------------------------------------------------+
| closed        | The issue has been resolved (somehow).                     |
+---------------+------------------------------------------------------------+

Linked pull requests
--------------------
A link might be added manually using the cog icon next to this field.
Most commonly though, if the PR includes "Fixes #xxx" in its description,

> the link will be added automatically.

Author: These sections have been removed.

+-------------------------------------------------------------+-------------------------------------------------------+
| Comment abbreviation                                        | Description                                           |
+=============================================================+=======================================================+
| ``#<number>``,                                              | Links to the tracker issue or PR ``<number>`` (they   |
| ``GH-<number>``                                             | share the same sequence of integers on GitHub).       |
+-------------------------------------------------------------+-------------------------------------------------------+
| ``BPO-<number>``                                            | Links to the old bug tracker at bugs.python.org.      |
+-------------------------------------------------------------+-------------------------------------------------------+
| a 10-, 11-, 12-, or 40-digit hex ``<number>``               | Indicates a Git changeset identifier and              |
|                                                             | generates a link to changeset ``<number>`` on GitHub. |

> +-------------------------------------------------------------+-------------------------------------------------------+

Author: This section has been adapted and moved to `issue-tracker.rst`.

>   state.  It also prevents ``miss-islington`` from being able to automatically

Reviewer 2: Not sure if we need code formatting on `miss-islington`

It's written "miss-islington" or "[@miss-islington](https://github.com/miss-islington)" at https://github.com/python/miss-islington

and "Miss Islington (bot)" at https://github.com/python/miss-islington

Shall we also link to the repo or user for one or more of these?

Reviewer: Yeah, I'd suggest

```suggestion
  state.  It also prevents `miss-islington <https://github.com/python/miss-islington>`
  from being able to automatically
```

Reviewer: This was implemented in the other comment, which reflowed it as well

>   state.  It also prevents ``miss-islington`` from being able to automatically
>   merge the PR.

Reviewer: Or, with more reasonable wrappping:

```suggestion
  state.  It also prevents `miss-islington <https://github.com/python/miss-islington>`
  from being able to automatically merge the PR.
```

>   indicate the stage of a PR:

Author: Perhaps we should link to the `bedevere` repo and keep the documentation of these labels there instead of here, since it's also subject to change.

>   indicate the stage of a PR:

Reviewer: Yeah, that could be a better idea :+1:

>   indicate the stage of a PR:

Author: I moved this last, removed the sub-list, and linked to https://github.com/python/bedevere/#pr-state-machine

>   a stuck build using the web interface.

Author: I just copied this as it was.  I agree the last sentence it's not very helpful, so maybe we can remove it?  Alternatively we could link to the buildbot page if it has more relevant info.

>   a stuck build using the web interface.

Reviewer: Okay, gotcha. At your discretion—I'm not that knowledgeable about the buildbots myself.

only to one. Below is a possibly inexhaustive list, but it should get
you going. For a full list, see `here <https://github.com/python/cpython/issues/labels>`_.

.. _general-purpose-labels:

> .. _Type:

Reviewer: ```suggestion
.. _Type:
.. _labels-type:
```

May as well add an actual ref target label here while we're at it, so people aren't tempted to use the legacy ref label or the ambiguous fragment-id only one.

Type labels
===========

These labels are used to specify the type of issue:

* :gh-label:`type-bug`: for unexpected behaviors, bugs, or exceptions
  (not hard crashes).
* :gh-label:`type-crash`: for hard crashes of the interpreter, possibly with a
  core dump.
* :gh-label:`type-feature`: for feature requests or enhancements.
  The `Ideas Discourse category`_ can be used to discuss enhancements
  before filing an issue.
* :gh-label:`type-security`: for security issues.
  See also `Reporting security issues in Python`_.

> .. _Component:

Reviewer: ```suggestion
.. _Component:
.. _labels-component:
```

Likewise

Component labels
================

These labels are mostly used to specify which :ref:`part of the codebase

> <directory-structure>` is affected by the issue/PR:

Reviewer: ```suggestion
<build-directory-structure>` is affected by the issue/PR:
```

Update to use disambiguated, non-title-duplicating ref target label

* :gh-label:`stdlib`: for standard library modules in the :cpy-file:`Lib`
  directory (written in Python).
* :gh-label:`extension-modules`: for standard library modules in the
  :cpy-file:`Modules` directory (written in C).
* :gh-label:`interpreter-core`: for changes related to the interpreter core in
  the :cpy-file:`Objects`, :cpy-file:`Python`, :cpy-file:`Grammar`,
  and :cpy-file:`Parser` dirs (written mostly in C).
* :gh-label:`docs`: for documentation in the :cpy-file:`Doc` directory
  (written in :ref:`reStructuredText <markup>`), docstrings, and code comments.
* :gh-label:`tests`: for tests in the :cpy-file:`Lib/test` directory

>   (written in Python) and other changes related to tests, ``unittest``,
>   or ``doctests``.

Reviewer: ```suggestion
  (written in Python) and other changes related to tests, :mod:`unittest`,
  or :mod:`doctests`.
```

* Cross-reference and format the relevant modules
* Be consistent about using either the module name literal vs. referring to the tests themselves in prose

Expert labels
=============

These labels are used to specify the area of expertise required to address
the issue/PR.  This includes both specific modules/packages and generic
interest areas.

Adding these labels is also a way to notify the relevant experts, since
they are encouraged to subscribe to them.  Depending on the label,
this might also automatically add the issue to a GitHub project.

You can see the `full list of expert labels on GitHub
<https://github.com/python/cpython/labels?q=expert>`_.

OS labels
=========

These labels are used to specify which operating systems are affected.
Since most issues either affect all systems or are specific to Unix,
the only available labels are :gh-label:`OS-windows`, :gh-label:`OS-mac`,
and :gh-label:`OS-freebsd`.

Version labels
==============

These labels are used to indicate which versions of Python are affected.
The available version labels (with the form :samp:`3.{N}`) are updated
whenever new major releases are created or retired.

See also :ref:`the branch status page <branchstatus>`
for a list of active branches.

> .. _Keywords:
> .. _Other:
> .. _Priority:

Reviewer: ```suggestion
.. _Keywords:
.. _Other:
.. _Priority:
.. _labels-other:
```

* Add an actual non-legacy, non-ambiguous ref target label here to use elsewhere

To note, the actual existing section-based fragment ids are lowercase, but ref target labels get lowercased too, and this is a reminder that these are there merely for backward compat to generate the appropriate fragment ids, and shouldn't be used and rather replaced with actual fragment-id based redirects in the future (which I'm working on).

Other labels
============

* :gh-label:`triaged`: for issue has been accepted as valid by a triager.
* :gh-label:`easy`: for issues that are considered easy.
* :gh-label:`build`/:gh-label:`performance`: for issues related respectively

>   to the build process and performances.

Reviewer: ```suggestion
  to the build process or performance, respectively.
```

* Fix grammar error
* Use phrasing appropriate to the fact that these are two separate classes of labels with two separate respective applications

Author: There's already a "respectively" on the previous line, do you mean to remove that?

Reviewer: Yes, sorry—my mistake!

The actually grammar error here wasn't actually fixed in the manual application, so I'll submit a followup suggestion and mark this one as resolved.

* :gh-label:`release-blocker`/:gh-label:`deferred-blocker`: for issues/PRs
  that, unless fixed, will hold the current or next release respectively.
  Triagers may set these labels, and the :ref:`branch's release manager <branchstatus>`
  will review them and determine if they indeed qualify,
  removing or retaining the label as appropriate.
* :gh-label:`pending`: for issues/PRs that will be closed unless further
  feedback is provided.
* :gh-label:`stale`: for issues/PRs that have been inactive for a while.

> * :gh-label:`sprint`: used for easier filtering of issues/PRs being worked on

Reviewer: ```suggestion
* :gh-label:`sprint`: for easier filtering of issues/PRs being worked on
```

To be consistent with all the others, which all start with "for"

  during official sprints.

> .. _GitHub Labels for PRs:

Reviewer: ```suggestion
.. _GitHub Labels for PRs:
.. _github-pr-labels:
```

Add back existing ref target label

Labels specific to PRs
======================

> The following labels only apply to PRs.  They are either set automatically

Reviewer: ```suggestion
The following labels only apply to :ref:`Pull Requests <pullrequest>`.  They are either set automatically
```

* Use the full term for clarity instead of repeating the acronym
* Link the appropriate page

interest areas.

Adding these labels is also a way to notify the relevant experts, since
they are encouraged to subscribe to them.  Depending on the label,
this might also automatically add the issue to a GitHub project.

You can see the `full list of expert labels on GitHub
<https://github.com/python/cpython/labels?q=expert>`_.

OS labels
=========

These labels are used to specify which operating systems are affected.
Since most issues either affect all systems or are specific to Unix,
the only available labels are :gh-label:`OS-windows`, :gh-label:`OS-mac`,
and :gh-label:`OS-freebsd`.

Version labels
==============

These labels are used to indicate which versions of Python are affected.
The available version labels (with the form :samp:`3.{N}`) are updated
whenever new major releases are created or retired.

See also :ref:`the branch status page <branchstatus>`
for a list of active branches.

.. _Keywords:
.. _Other:
.. _Priority:
.. _labels-other:

Other labels
============

* :gh-label:`triaged`: for issue has been accepted as valid by a triager.
* :gh-label:`easy`: for issues that are considered easy.
* :gh-label:`build`/:gh-label:`performance`: for issues related

>   to the build process or performances, respectively.

Reviewer: ```suggestion
  to the build process or performance, respectively.
```

Fix grammar error (per previous suggestion)

only to one. Below is a possibly inexhaustive list, but it should get
you going. For a full list, see `here <https://github.com/python/cpython/issues/labels>`_.

> Type labels

Author: ```suggestion
.. _general-purpose-labels:
.. _Type:

Type labels
```

This replaces the other two suggestions, that are conflicting because they edit the same lines.
